core/src/main/java/org/apache/iceberg/TableMetadata.java

```diff
@@ -1047,17 +725,543 @@ private static SortOrder freshSortOrder(int orderId, Schema schema, SortOrder so
+      if (order.isUnsorted()) {
+        newOrder = SortOrder.unsorted();
+      } else {
+        // rebuild the sort order using new column ids
+        newOrder = freshSortOrder(newOrderId, schema, order);
+      }
+
+      sortOrders.add(newOrder);
+      sortOrdersById.put(newOrderId, newOrder);
+
+      changes.add(new MetadataUpdate.AddSortOrder(newOrder));
+
+      return newOrderId;
+    }
+
+    private int reuseOrCreateNewSortOrderId(SortOrder newOrder) {
+      if (newOrder.isUnsorted()) {
+        return SortOrder.unsorted().orderId();
+      }
+
+      // determine the next order id
+      int newOrderId = INITIAL_SORT_ORDER_ID;
+      for (SortOrder order : sortOrders) {
+        if (order.sameOrder(newOrder)) {
+          return order.orderId();
+        } else if (newOrderId <= order.orderId()) {
+          newOrderId = order.orderId() + 1;
+        }
+      }
+
+      return newOrderId;
+    }
+
+    private void setCurrentSnapshot(Snapshot snapshot, long currentTimestampMillis) {
+      if (snapshot == null) {
+        this.currentSnapshotId = -1;
+        snapshotLog.clear();
+        return;
+      }
+
+      if (currentSnapshotId == snapshot.snapshotId()) {
+        return;
+      }
+
+      ValidationException.check(formatVersion == 1 || snapshot.sequenceNumber() <= lastSequenceNumber,
+          "Last sequence number %s is less than existing snapshot sequence number %s",
+          lastSequenceNumber, snapshot.sequenceNumber());
+
+      this.lastUpdatedMillis = currentTimestampMillis;
+      this.currentSnapshotId = snapshot.snapshotId();
+      snapshotLog.add(new SnapshotLogEntry(lastUpdatedMillis, snapshot.snapshotId()));
+      changes.add(new MetadataUpdate.SetCurrentSnapshot(snapshot.snapshotId()));
+    }
+
+    private List<MetadataLogEntry> addPreviousFile(String previousFileLocation, long timestampMillis) {
+      if (previousFileLocation == null) {
+        return previousFiles;
+      }
+
+      int maxSize = Math.max(1, PropertyUtil.propertyAsInt(properties,
+          TableProperties.METADATA_PREVIOUS_VERSIONS_MAX, TableProperties.METADATA_PREVIOUS_VERSIONS_MAX_DEFAULT));
+
+      List<MetadataLogEntry> newMetadataLog;
+      if (previousFiles.size() >= maxSize) {
+        int removeIndex = previousFiles.size() - maxSize + 1;
+        newMetadataLog = Lists.newArrayList(previousFiles.subList(removeIndex, previousFiles.size()));
+      } else {
+        newMetadataLog = Lists.newArrayList(previousFiles);
+      }
+      newMetadataLog.add(new MetadataLogEntry(timestampMillis, previousFileLocation));
+
+      return newMetadataLog;
+    }
+
+    /**
+     * Finds intermediate snapshots that have not been committed as the current snapshot.
+     *
+     * @return a set of snapshot ids for all added snapshots that were later replaced as the current snapshot in changes
+     */
+    private static Set<Long> intermediateSnapshotIdSet(List<MetadataUpdate> changes, long currentSnapshotId) {
```

Comment:
This is new. It identifies intermediate snapshots that were added and set as current without being committed. The logs for these are suppressed.

Reply:
Is it necessary to check the entire list of changes instead of just the changes after the `startingChangeCount`? It seems that `updateSnapshotLog` does not affect snapshots that are already committed.